[Ready] SSthrowing + callbacks!

MrsTonedOne opened the pull request:
- Throwing is now a subsystem.
- It's low priority, but is a ticker subsystem so is ran before most other subsystems.
- To allow for shit to run after the throw finishes, throw_at now supports a callback. (since it can no longer block)
- **A callback datum system was created**, conversion of addtimer is planned for another PR.
- Throwing now has a limit of 512 turfs (was 600)
- Throwing now ticks every world.tick, (it converts the speed arg from turfs per 1ds to turfs per what ever tick_lag is) (under default fps, that means throwing now ticks twice as often as the old method, moving the thing half as many turfs each tick)
- Throwing now properly accounts for missed ticks.
- Throwing no longer uses sleep.
- Throwing should no longer lag since it's not filling the sleep queue up
- :new: Item throwing now imparts the momentum of the user throwing. (this is to make hitting yourself with thing you are throwing much less likely)

:cl: MrStonedOne
tweak: Throwing was refactored to cause less lag and be more precise
add: Item throwing now imparts the momentum of the user throwing. Throwing in the direction you are moving throws the item faster, throwing away from the direction you are moving throws the item slower. This should make hitting yourself with floor tiles less likely.
/:cl:

I don't know how this fits with the freeze, but i'm calling it a refactor.

fixes #19187
fixes #20305
(mostly fixes) #18670 (it can still happen in certain edge cases because movement doesn't follow normal byond ticks and can in fact happen in between them, but i'm calling that a feature since the likelihood of it happening increases the slower the item is thrown, hitting yourself with a paper when throwing it is just hilarious)

MrsTonedOne committed:
SSthrowing + callbacks!
Throwing is now a subsystem.
It's low priority, but is a ticker subsystem so is ran before most other subsystems.
To allow for shit to run after the throw finishes, throwing now supports a callback.
A callback datum system was created, conversion of addtimer is planned for another PR.
Throwing now has a limit of 2048 turfs (was 600)
Throwing now ticks every world.tick, and properly converts the speed arg from 1ds to what ever tick_lag is.
Throwing now properly accounts for missed ticks.
Throwing no longer uses sleep.
Throwing should no longer lag since it's not filling the sleep queue up

MrStonedOne commented on code/datums/callback.dm:
```diff
@@ -0,0 +1,90 @@
+		var/datum/callback/C = new(object|null, /proc/type/path|"procstring", arg1, arg2, ... argn)
+		var/timerid = addtimer(C, time, timertype)
+		OR
+		var/timerid = addtimer(CALLBACK(object|null, /proc/type/path|procstring, arg1, arg2, ... argn), time, timertype)
```
these don't work yet since addtimer hasn't been converted, but they will.

MrStonedOne commented on code/datums/callback.dm:
```diff
@@ -0,0 +1,90 @@
+		proc defined on current(src) object (when in a /proc/ and not an override) OR overridden at src or any of it's parents:
+			.procname
+			Example:
+				CALLBACK(src, .some_proc_here)
```
@Fox-McCloud this might interest you, since you were to first to sound the alarm over abusing string procs

XDTM commented:
Can i run code at the end of the throw now? If so this would make tentacles work more smoothly

MrStonedOne commented:
you always could. it was only throw_at_fast that didn't support that(because all it was was throw_at but didn't block). throw_at was normally blocking, as in it didn't return until the throw finished.

Now it's back to non-blocking.

Anywho, for your tentacles i wasn't sure if the addtimer was a delay, or if you thought throw_at was non-blocking, so i kept it, i'll change that over to just call the grab/stab proc directly

MrsTonedOne committed:
Smoother tentacles

XDTM commented:
Thanks

MrsTonedOne committed:
Some improvements

PKPenguin321 commented:
Why is MSO so based?

optimumtact commented on code/datums/martial/wrestling.dm:
```diff
@@ -182,9 +182,8 @@
 				D.emote("scream")
-			D.throw_at(T, 10, 4)
+			D.throw_at(T, 10, 4, callback = CALLBACK(D, /mob/living/carbon/human/.Weaken, 2))
 			D.Weaken(2)
```
won't this weaken them, then weaken them again?

MrStonedOne replied:
ya, i forgot to remove the line once i made it a callback

Iamgoofball commented:
What's a callback and how does it work in mememan's terms

RemieRichards commented:
@Iamgoofball the name is the clue, "callback"

"Can I give you a call back in 5 minutes?"
"Sure"

is the same as
Procname() Called
Procname() Running
Procname() Finishes
Callbackname() Called automatically

MrStonedOne commented:
http://www.byond.com/forum/?post=2191707 This is one of those things that's easier to explain with code then with words. it's just a wrapper around procs that take procs to call as arguments. the thingtocall, proctocall, args.... mess that addtimer has, that sorttim has, etc. this is a generic wrapper around that concept.

Lets say you want `/mob/living/proc/after_item_throw` to be called after a throw mode throw finishes, Since throwing is a subsystem, you can't just call `throw_at`, then call `after_item_throw`, so you have to tell `throw_at` to call it when the throw finishes. with this system, you just do:
````
/mob/living/proc/click_on_tile(some shit here)
	if (throwmode)
		item.throw_at(thing_clicked,some_other_shit, callback = CALLBACK(src, .proc/after_item_throw, item, some, other, args, to, call, after_item_throw, with.)
````

Now, all `throw_at` has to do, is store the created callback datum, and run `callback`.`Invoke()` when the throw finishes to run after_item_throw

a callback is just a reference to a proc call and args that's easier to use and work with.

Iamgoofball commented:
💯🍆👌👌👌👌👌👌👌👌👌👌👌💯💯💯<😘😘

Iamgoofball commented:
I wonder if we can use this for some Tesla stuff

RemieRichards commented:
@MrStonedOne 

>easier in code than in words
>literally explained in 2 words

uh huh.

MrStonedOne commented:
>I wonder if we can use this for some Tesla stuff

Your mind is weird conceptually.

Explain to me how you arrived at that conclusion. where you read about that, and thought tesla.

MrsTonedOne committed:
Missed a spot.

Iamgoofball commented:
right so we might be able to abuse callbacks to clean up tesla_zap, I'll explain better when I'm at a computer

MrsTonedOne committed:
Makes shit quicker.
Inlines the thrownthing.tick() proc.
Raises missed ticks value
Lowers max dist value
Inlines the two sister overrides for /atom/movable/Moved() because that just seemed like a waste

optimumtact commented:
some wack ass conflitcs going on here

MrsTonedOne committed:
>PRs open that use procs i'm removing.

MrStonedOne commented:
Test merging on bagil

MrsTonedOne committed:
STOP THE PRESSES!

AnturK commented:
Is this finished ?

MrStonedOne commented:
oh. uhh, I was trying to fix the issue where throwing while running caused you to hit yourself. i'll go re-checkout that branch and see where i was on that.

MrsTonedOne committed:
throw_at now runs the first throw tick() immediately
This will help some with throwing while running.

MrsTonedOne committed:
Item throwing now imparts the momentum of the user throwing.

(ie, running in the direction you are throwing makes you throw faster, running away from the direction you are throwing makes you throw the item slower)

MrsTonedOne committed:
Moves throwing momentum from carbon/throw_item to movable/throw_at.
There are other things that cause a mob to "throw" an item, I figured we keep this universal since thrower is already an arg.

MrStonedOne commented:
I'm gonna rebase then re-check for anything merged recently that depended on old behavior, then this can be merged.

MrsTonedOne committed:
Explosions throw shit faster.
This was stupid, "Hey, lets set the item's throw_speed to 4 so embedding works, but lets make it throw at the base 2 throw speed for no reason."

MrsTonedOne committed:
Fixes explosion embedding.
This also acts as a nice example of how to override a callback in an override of throw_at properly.
